### PR TITLE
Add UnmanagedCallConvAttribute

### LINKED ISSUE #51156: Extensible Calling Conventions for native callee functions
## Background and Motivation

The [`UnmanagedCallersOnlyAttribute`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.unmanagedcallersonlyattribute.callconvs) provides an extensible mechanism to define native calling conventions on a .NET method. This enables users to author functions that are callable in a native environment. A symmetrical need arises for functions that are defined in native code but declared to be called in a .NET environment. These native function declarations are done using [`DllImportAttribute`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute).

This approach complements `UnmanagedCallersOnlyAttribute` and aligns with the reuse of the `CallConv*` types under the [`System.Runtime.CompilerServices`](https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices) namespace.

## Proposed API

If the user specifies the `CallingConvention.Winapi` (the default) on the `DllImportAttribute` the following attribute will be queried.

```diff
namespace System.Runtime.InteropServices
{
+    /// <summary>
+    /// Provides an equivalent to <see cref="UnmanagedCallersOnlyAttribute"/> for native
+    /// functions declared in .NET.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
+    public class UnmanagedCalleeAttribute : Attribute
+    {
+        public UnmanagedCalleeAttribute() { }
+
+        /// <summary>
+        /// Types indicating calling conventions for the unmanaged callee.
+        /// </summary>
+        /// <remarks>
+        /// If <c>null</c>, the semantics are identical to <c>CallingConvention.Winapi</c>.
+        /// </remarks>
+        public Type[]? CallConvs;
+    }
}
```

## Usage Examples

**P/Invoke**

Using the defaults, all 3 of the following declarations are the same.
```csharp
[DllImport("nativelib")]
static extern int Func(IntPtr a, nint val);

[DllImport("nativelib", CallingConvention = CallingConvention.Winapi)]
static extern int Func(IntPtr a, nint val);

[UnmanagedCallee]
[DllImport("nativelib")]
static extern int Func(IntPtr a, nint val);
```

Some calling conventions can be declared in multiple ways.
```csharp
[DllImport("nativelib", CallingConvention = CallingConvention.Cdecl)]
static extern int Func(IntPtr a, nint val);

[UnmanagedCallee(CallConvs = new[] { typeof(CallConvCdecl) })]
[DllImport("nativelib")]
static extern int Func(IntPtr a, nint val);

[UnmanagedCallee(CallConvs = new[] { typeof(CallConvCdecl) })]
[DllImport("nativelib", CallingConvention = CallingConvention.Winapi)]
static extern int Func(IntPtr a, nint val);
```

Future calling conventions would be enabled without updating the [`CallingConvention`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.dllimportattribute.callingconvention) enum. Examples in C# and VB.NET.
```csharp
[UnmanagedCallee(CallConvs = new[] { typeof(CallConvVectorCall) })]
[DllImport("nativelib")]
static extern int Func(IntPtr a, nint val);
```

```vbnet
<UnmanagedCallee(CallConvs := New Type() { GetType(CallConvVectorCall) })>
<DllImport("nativelib")>
Public Shared Function Func(a As IntPtr, val As IntPtr) As Integer
End Function
```

This proposal provides an alternative mechanism to support Native Varargs in https://github.com/dotnet/runtime/issues/48796. 

**COM style interface** - Not proposed but example of how the attribute could evolve.

To enable this scenario the attribute would be updated with the following: `[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]`

Using the defaults, all 3 of the following declarations are the same on Windows.
```csharp
[ComVisible(true)]
[Guid("C509A1C1-A6C9-4229-89CE-10B53CF8271B")]
interface IFoo
{
    int Call(ref SomeType valuetype);
}

[ComVisible(true)]
[Guid("C509A1C1-A6C9-4229-89CE-10B53CF8271B")]
[UnmanagedCallee]
interface IFoo
{
    int Call(ref SomeType valuetype);
}

[ComVisible(true)]
[Guid("C509A1C1-A6C9-4229-89CE-10B53CF8271B")]
[UnmanagedCallee(CallConvs = new[] {  typeof(CallConvStdCall), typeof(CallConvMemberInstance) })]
interface IFoo
{
    int Call(ref SomeType valuetype);
}
```

## Alternative Designs

Provide an enum to trigger off for reading the new attribute.

```diff
namespace System.Runtime.InteropServices
{
    public enum CallingConvention
    {
+        /// <summary>
+        /// Instructs the runtime to consult the <see cref="UnmanagedCalleeAttribute"/>.
+        /// </summary>
+        UnmanagedCallee = 7
    }
}
```

Alternative name proposed could be `CallingConvention.UnmanagedCallConv`.

### PR DESCRIPTION
**Pull request by elinor-fung**
Fixes https://github.com/dotnet/runtime/issues/51156

- Add `UnmanagedCallConvAttribute`
- Use `UnmanagedCallConv` if the calling convention is not specified or specified as `winapi` on pinvokeimpl metadata
  - Also used for suppressing GC transition, since `CallConvSuppressGCTransition` can be specified
  - Check attribute in mono - cc @CoffeeFlux @lambdageek
  - Check attribute in crossgen2 - cc @trylek 
- Add tests
- Not handled in crossgen
  - Similar to `UnmanagedCallersOnly` - since we don't include the support to read a type array, make crossgen bail if the method is marked `UnmanagedCallConv`
 
This does not implement checking of `UnmanagedCallConv` for delegates.

cc @AaronRobinsonMSFT @jkoritzinsky

### CODE REVIEW

**Comment by jkotas** [comment]
> This also does not implement checking of UnmanagedCallConv for delegates

I would not bother with implementing the advanced calling conventions for delegates. I think we should make everybody who needs to do interop with function pointers to use function pointers.

File: src/mono/mono/metadata/marshal.c [hunk 1]
```diff
@@ -3069,19 +3089,65 @@ mono_marshal_set_callconv_from_modopt (MonoMethod *method, MonoMethodSignature *
...
-				csig->call_convention = MONO_CALL_FASTCALL;
-			else if (!strcmp (cmod_class_name, "CallConvThiscall"))
-				csig->call_convention = MONO_CALL_THISCALL;
+		mono_marshal_set_callconv_for_type (cmod_type, csig, NULL);
+	}
+}
+
+static void
+mono_marshal_set_callconv_from_unmanaged_callconv_attribute (MonoMethod *method, MonoMethodSignature *csig, gboolean *skip_gc_trans /*out*/)
+{
```
Comment (lambdageek) [inline src/mono/mono/metadata/marshal.c:3098]: A couple of general comments about this method:
1. As written now, it's not usable by the AOT compiler.
2. Can you use `mono_reflection_create_custom_attr_data_args_noalloc` - that should avoid creating managed objects and will instead just give you C structures.  (It's possible it doesn't handle `Type[]` yet, in which case we should fix that - as a separate PR.)
3. There's a pattern to cleanup `MonoCustomAttrInfo` values:
   ```
      	if (!cinfo->cached)
		mono_custom_attrs_free (cinfo);
   ```
4. If the managed methods can't be avoided, the portion of the function that uses handles should be:
  a. pulled out into its own function, 
  b. that function should be single-exit and the body of it should be wrapped in `HANDLE_FUNCTION_ENTER()`/`HANDLE_FUNCTION_RETURN_{,VAL,OBJ}()`

**Review by lambdageek — COMMENTED** [review]
It would be great if the mono code could avoid managed allocation to decode the custom attribute.  That would make it usable for AOT compilation.  If that turns out not to work, we should still take the code, disabled, and do a follow-up fix to `mono_reflection_create_custom_attr_data_args_noalloc` to make it work.

File: src/mono/mono/metadata/marshal.c (hunk 1 above)
Comment (elinor-fung) [reply src/mono/mono/metadata/marshal.c:3098]: > It's possible it doesn't handle `Type[]`

This is what I hit when I had tried to use the noalloc version. `mono_reflection_create_custom_attr_data_args_noalloc` doesn't handle arrays - `load_cattr_value` requires being given an out obj for arrays.

Comment (lambdageek) [reply src/mono/mono/metadata/marshal.c:3098]: I see.  well we should take the managed object version (with the other two cleanups I mentioned) then, and open a GH Issue to improve `_noalloc` and get the AOT compiler support in.

**Comment by AaronRobinsonMSFT** [comment]
>I would not bother with implementing the advanced calling conventions for delegates.

I am okay with this also. I would like to see us error out though if it is set on a `Delegate` - unless someone can argue it being a no-op is more appropriate.

**Comment by jkotas** [comment]
> I would like to see us error out though if it is set on a Delegate

If you try to use this on a Delegate, the compiler will produce an error thanks to  `AttributeTargets.Method`.

File: src/mono/mono/metadata/marshal.c (hunk 1 above)
Comment (elinor-fung) [reply src/mono/mono/metadata/marshal.c:3116]: Filed https://github.com/dotnet/runtime/issues/52977

**Review by lambdageek — COMMENTED** [review]
mono bits lgtm

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/coreclr/vm/jitinterface.cpp [hunk 2]
```diff
@@ -9997,12 +9997,15 @@ namespace
...
             if (pMD->IsNDirect())
             {
-                if (pSuppressGCTransition)
-                {
-                    *pSuppressGCTransition = pMD->ShouldSuppressGCTransition();
-                }
+#ifdef CROSSGEN_COMPILE
+                // Return CorInfoCallConvExtension::Managed to indicate that crossgen does not support handling UnmanagedCallConv
+                if (pMD->HasUnmanagedCallConvAttribute())
+                    return CorInfoCallConvExtension::Managed;
```
Comment (jkoritzinsky) [inline src/coreclr/vm/jitinterface.cpp:10003]: Does the JIT handle the case when the "Managed" callconv is returned from here correctly? (It should decide that it cannot inline the call as an unmanaged call).

Comment (elinor-fung) [reply src/coreclr/vm/jitinterface.cpp:10003]: Yeah, `impCheckForPInvokeCall` checks for `CorInfoCallConvExtension::Managed` as indication that the unmanaged calling convention couldn't be determined and bails without inlining the call.

**Review by jkoritzinsky — APPROVED** [review]

**Comment by elinor-fung** [comment]
All the tests passed - failure was from uploading results (`azure.devops.exceptions.AzureDevOpsClientRequestError: Operation returned an invalid status code of 503`)

**Pull request merged**